services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java

```diff
@@ -433,11 +436,21 @@ protected Response handleLoginResponse(String samlResponse, SAMLDocumentHolder h
                     return ErrorPage.error(session, authSession, Response.Status.BAD_REQUEST, Messages.INVALID_REQUESTER);
                 }
 
-                Element assertionElement;
+                Element assertionElement = null;
 
                 if (assertionIsEncrypted) {
-                    // This methods writes the parsed and decrypted assertion back on the responseType parameter:
-                    assertionElement = AssertionUtil.decryptAssertion(holder, responseType, keys.getPrivateKey());
+                    try {
+                        /* This code is deprecated and will be removed in Keycloak 24 */
+                        if (DEPRECATED_ENCRYPTION) {
+                            KeyManager.ActiveRsaKey keys = session.keys().getActiveRsaKey(realm);
+                            assertionElement = AssertionUtil.decryptAssertion(responseType, keys.getPrivateKey());
+                        } else {
+                        /* End of deprecated code */
+                            assertionElement = AssertionUtil.decryptAssertion(responseType, new SAMLIDPDecryptionKeysLocator(session, realm, config));
+                        }
+                    } catch (ProcessingException ex) {
+                        throw new WebApplicationException(ex, Response.Status.BAD_REQUEST);
```

Comment:
Is it possible to display some useful error message to the server.log for this case?

I guess that people migrating from previous version and relying on previous behaviour may have issues and may not know why things not working. I can imagine some warning like `Not possible to decrypt SAML assertion. Please check realm keys of usage ENC in the realm 'foo' and make sure there is a key able to decrypt the assertion encrypted by identity provider 'my-saml-idp'` . WDYT?

Feel free to ignore this comment if you think that error message is already useful, but my guess is that not (actually not 100% sure if `WebApplicationException` displays something to the server.log at all...)

Reply:
I agree. Not only for previous false use case. Logs should show what is wrong with keys (fe false key configuration).